exit/test_exit.c

```diff
@@ -0,0 +1,811 @@
+} test_common;
+
+
+/* SIGCHLD signal handler */
+static void test_sigchldHandler(int signum)
+{
+	struct sigaction sa;
+
+	sa.sa_handler = SIG_DFL;
+	TEST_ASSERT_EQUAL_INT(0, sigaction(SIGCHLD, &sa, NULL));
+
+	/* Change value of variable to confirm that handler has been invoked */
+	test_common.test_handlerFlag = TEST_EXIT_DUMMY_VAL;
+}
+
+
+static void test_sigusrHandler(int signum)
+{
+	struct sigaction sa;
+
+	sa.sa_handler = SIG_IGN;
+	sigaction(SIGUSR1, &sa, NULL);
+}
+
+
+static void test_destructor(void *arg)
+{
+	creat(TEST_EXIT_PATH, DEFFILEMODE);
+}
+
+
+static void *test_threadFunc(void *arg)
+{
+	pthread_key_t key;
+
+	pthread_key_create(&key, test_destructor);
+	pthread_setspecific(key, arg);
+
+	while (!test_common.test_threadWait) {
+		;
+	}
+	test_common.test_exitPtr(0);
+}
+
+
+/* Functions to be registered by atexit */
+static void test_atexit_fun1(void)
+{
+	int fd, val = 1;
+
+	fd = open(TEST_EXIT_PATH, O_WRONLY | O_APPEND);
+	write(fd, &val, sizeof(int));
+	close(fd);
+}
+
+
+static void test_atexit_fun2(void)
+{
+	int fd, val = 1234;
+
+	fd = open(TEST_EXIT_PATH, O_WRONLY | O_APPEND);
+	write(fd, &val, sizeof(int));
+	close(fd);
+}
+
+
+static void test_no_atexit(void)
+{
+	creat(TEST_EXIT_PATH, DEFFILEMODE);
+}
+
+
+TEST_GROUP(unistd_exit);
+
+
+TEST_SETUP(unistd_exit)
+{
+	remove(TEST_EXIT_PATH);
+}
+
+
+TEST_TEAR_DOWN(unistd_exit)
+{
+}
+
+
+TEST(unistd_exit, status_vals)
+{
+	/* Return all possible least significant byte values and check whether wait() works */
+	pid_t pid;
+	int i;
+	int val[3];
+	/*
+	 * Only the least significant 8 bits shall be available to a waiting parent process,
+	 * 3 most significant bytes should be cut off so returning val[0] should resolve to returning 0, val[1] to 1 and val[2] to 2
+	 */
+	val[0] = 0x1 << 8;
+	val[1] = (0x1 << 16) + 1;
+	val[2] = (0x1 << 24) + 2;
+
+	for (i = 0; i < 256; i++) {
+		pid = fork();
+		TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+		/* child */
+		if (pid == 0) {
+			test_common.test_exitPtr(i);
+		}
+		/* parent */
+		else {
+			int status, ret;
+
+			ret = wait(&status);
+			TEST_ASSERT_EQUAL_INT(pid, ret);
+			TEST_ASSERT_EQUAL_INT(i, WEXITSTATUS(status));
+		}
+	}
+	/* Some exceeding values */
+	for (i = 0; i < 3; i++) {
+		pid = fork();
+		TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+		/* child */
+		if (pid == 0) {
+			test_common.test_exitPtr(val[i]);
+		}
+		/* parent */
+		else {
+			int status, ret;
+
+			ret = wait(&status);
+			TEST_ASSERT_EQUAL_INT(pid, ret);
+			TEST_ASSERT_EQUAL_INT(i, WEXITSTATUS(status));
+		}
+	}
+}
+
+
+TEST(unistd_exit, exit_status_waitpid)
+{
+	/* Check if after _exit() waitpid() works */
+	pid_t pid;
+
+	pid = fork();
+	TEST_ASSERT_GREATER_OR_EQUAL(0, pid);
+	/* child */
+	if (pid == 0) {
+		test_common.test_exitPtr(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		int status, ret;
+
+		ret = waitpid(pid, &status, 0);
+		TEST_ASSERT_EQUAL_INT(pid, ret);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+	}
```

Comment:
We could create a case, where a few threads would be created in parent process. They could all wait using waitpid/wait (it could be good to use both) and write status to a file if a status has been read. Finally only one status should be written.
![Screenshot from 2023-09-29 11-08-00](https://github.com/phoenix-rtos/phoenix-rtos-tests/assets/77304431/e26f4cb1-1f2f-42d3-8d75-85813a5b25cb)